File: Linear Regression/Algorithms in C++/linearregression.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

double calculate_mean(vector<double> data) {
    double sum = 0;
    for (double d : data) {
        sum += d;
    }
    return sum / data.size();
}

double calculate_variance(vector<double> data) {
    double mean = calculate_mean(data);
    double variance = 0;
    for (double d : data) {
        variance += pow(d - mean, 2);
    }
    return variance / data.size();
}

double calculate_covariance(vector<double> x, vector<double> y) {
    double x_mean = calculate_mean(x);
    double y_mean = calculate_mean(y);
    double covariance = 0;
    for (int i = 0; i < x.size(); i++) {
        covariance += (x[i] - x_mean) * (y[i] - y_mean);
    }
    return covariance / x.size();
}

void linear_regression(vector<double> x, vector<double> y, double& slope, double& intercept) {
    double x_variance = calculate_variance(x);
    double y_variance = calculate_variance(y);
    double covariance = calculate_covariance(x, y);

    slope = covariance / x_variance;
    intercept = calculate_mean(y) - slope * calculate_mean(x);
}

int main() {
    vector<double> x = {1, 2, 3, 4, 5};
    vector<double> y = {2, 4, 6, 8, 10};

    double slope, intercept;
    linear_regression(x, y, slope, intercept);

    cout << "The equation of the line of best fit is y = " << slope << "x + " << intercept << endl;

    return 0;
}

File: Linear Regression/Algorithms in C#/linearregression.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LinearRegressionExample
{
    class Program
    {
        static void Main(string[] args)
        {
            List<double> x = new List<double>() { 1, 2, 3, 4, 5 };
            List<double> y = new List<double>() { 2, 4, 6, 8, 10 };

            double slope, intercept;
            LinearRegression(x, y, out slope, out intercept);

            Console.WriteLine("The equation of the line of best fit is y = {0}x + {1}", slope, intercept);

            Console.ReadLine();
        }

        static void LinearRegression(List<double> x, List<double> y, out double slope, out double intercept)
        {
            double x_mean = x.Average();
            double y_mean = y.Average();

            double covariance = x.Zip(y, (xi, yi) => (xi - x_mean) * (yi - y_mean)).Sum();

            double x_variance = x.Sum(xi => Math.Pow(xi - x_mean, 2));
            double y_variance = y.Sum(yi => Math.Pow(yi - y_mean, 2));

            slope = covariance / x_variance;
            intercept = y_mean - slope * x_mean;
        }
    }
}
